### PR TITLE
Add interface structure

### PR DESCRIPTION
**Pull request by kmdeck**
PR adding in some structure to setting the RHS, initial conditions, and boundary conditions for the soil model.

Currently this is written such that prescribed models are still stored in the state vector.

Note that 2 of the tests do not use this interface yet, because I need to update the BC to handle dirichlet conditions to do so, but one does (hydrology only), and an experiment does (full model).

still to come: simulation structure, passing in domain info, further BC work, parameter work, having the simulation initialize the state vector to a default.

### CODE REVIEW

File: src/SoilModel/right_hand_side.jl [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+function make_rhs(model::SoilModel)
+    rhs_soil! = make_rhs!(model.energy_model, model.hydrology_model, model)
+    function rhs!( dY, Y, p, t)
+        rhs_soil!(dY,Y,p,t) 
+        return dY
+    end
+    return rhs!
+end
+
+function make_rhs!(energy::PrescribedTemperatureModel, hydrology::PrescribedHydrologyModel, model)
```
Comment (kmdeck) [inline src/SoilModel/right_hand_side.jl:12]: nominally, need 4 methods, as soil energy and soil hydrology each have two types.
We could do 4 methods (A) - soil_energy_rhs! x2 and soil_hydrology_rhs! x2, and then
compute rhs_soil! from applying both in series or
(B) soil_rhs(energy, hydrology) x 4.

We are choosing the latter (B) because (1) in the full dynamic model, the RHS for water and energy
are very intertwined, and treating them seperately (as in option A) requires doing calculations 2x and
(2) in the case where one is prescribed and the other is not, we need to worry about setting the prescribed
state to the correct value first (f(x,y,z,t_current)) before computing the RHS of the dynamic variable.
this ordering would change depending on which is prescribed.

File: src/SoilModel/initial_conditions.jl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    hydrology_ic::H
+    energy_ic::E
+end
+
+function SoilIC(;energy_ic = energy_ic, hydrology_ic = hydrology_ic)
+    args = (hydrology_ic, energy_ic)
+    return SoilIC{typeof.(args)...}(args...)
+end
+
+function create_initial_state(model::SoilModel, ic::SoilIC, zc)
```
Comment (kmdeck) [inline src/SoilModel/initial_conditions.jl:15]: The downside here is that with prescribed models  (f(x,y,z,t)), we are storing the values and calculating them/updating them (once) to allow for time variation. But then they are accessible by any module that needs them. In a different approach, we would just calculate them when needed and not store them, but this could lead to more calculations.

File: src/SoilModel/parameters.jl [hunk 3]
```diff
@@ -0,0 +1,25 @@
+export SoilParams
+
+abstract type ParameterStructure{FT <: AbstractFloat} end
+
+struct SoilParams{FT} <: ParameterStructure{FT}
```
Comment (kmdeck) [inline src/SoilModel/parameters.jl:5]: will have to clean up later - give defaults, etc.

File: src/SoilModel/right_hand_side.jl [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        dϑ_l = dY_hydro.ϑ_l
+        dθ_i = dY_hydro.θ_i
+
+        #update the state Y with prescribed values at the current time
+        ϑ_l = hydrology.ϑ_l_profile.(zc, t)
+        θ_i = hydrology.θ_i_profile.(zc, t)
+        θ_l = ϑ_l # eventually have a conversion to liquid water content
+        T = energy.T_profile.(zc, t)
+        ρc_s = volumetric_heat_capacity.(θ_l, θ_i, model.soil_param_set.ρc_ds, Ref(model.earth_param_set))
+        ρe_int = volumetric_internal_energy.(θ_i, ρc_s, T, Ref(model.earth_param_set))
```
Comment (kmdeck) [inline src/SoilModel/right_hand_side.jl:28]: Because we prescribe temperature, this requires a little extra step to convert to energy (energy is our state variable.). we then need to convert from energy back to T when T is needed...

File: src/SoilModel/right_hand_side.jl [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        top_water_flux, btm_water_flux = water_flux.top_flux, water_flux.btm_flux
+        sp = model.soil_param_set
+        param_set= model.earth_param_set
+        @unpack ν, vgn, vgα, vgm, ksat, θr, S_s, ρc_ds = sp
+        
+        # Evaluate T at the current time, update ρe_int
+        cs = axes(θ_i)
+        zc = Fields.coordinate_field(cs)
+        θ_l = ϑ_l
+        T = energy.T_profile.(zc, t)
```
Comment (kmdeck) [inline src/SoilModel/right_hand_side.jl:65]: not being used currently, but will be.

File: src/SoilModel/right_hand_side.jl [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        
+        @. dϑ_l = -gradf2c_water(-interpc2f(K) * gradc2f_water(h)) #Richards equation
+        dθ_i = Fields.zeros(eltype(θ_i), cs) # zero unless we add phase change
+        return dY
+    end
+    return rhs!
+end
+
+# Simple heat equation in soil. Water profile is prescribed. No heat flux due to diffusion of water.
+function make_rhs!(energy::SoilEnergyModel, hydrology::PrescribedHydrologyModel, model::SoilModel)
```
Comment (kmdeck) [inline src/SoilModel/right_hand_side.jl:101]: one sort of odd thing here is that the user could in principle supply water content of (z,t), but the change in water content with time (due to prescribed function) does not imply a water flux throughout the soil. Energy only changes due to gradient in T and thermal properties (fn of water content)

File: src/SoilModel/boundary_conditions.jl [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+struct NoBC <: AbstractBC end
+
+struct FluxBC{FT <: AbstractFloat} <: AbstractBC
+    top_flux::FT
+    btm_flux::FT
+end
+
+Base.@kwdef struct SoilBC{ebc<: AbstractBC, hbc<:AbstractBC}
+    energy_bc::ebc = NoBC()
```
Comment (kmdeck) [inline src/SoilModel/boundary_conditions.jl:12]: not actually being used currently, but it's nice to have a default for prescribed cases

File: src/SoilModel/initial_conditions.jl [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    Y_hydrology = ic.hydrology_ic.(zc, Ref(model))
+    Y_energy = ic.energy_ic.(zc, Ref(model))
```
Comment (kmdeck) [inline src/SoilModel/initial_conditions.jl:17]: if the user wants a prescribed water model, they have already specified water(x,y,z,t) in model.hydrology. Here we ask them to specify it again to put in the IC. The upside is that ic.hydrology_ic will work for both prescribed and dynamic cases, the downside is that the user has to specify code 2x. but on their end, in the driver file, they can write something like:
hydrology_ic = model.hydrology.theta_profile(x,y,z,t_initial) at least.

Comment (kmdeck) [reply src/SoilModel/initial_conditions.jl:17]: I think this is OK for now. asking the user to specify an IC for prescribed variables is a bit odd, but it is consistent with how they are stored internally (as part of Y).

Comment (kmdeck) [reply src/SoilModel/initial_conditions.jl]: yes, going to punt on this for now: https://github.com/CliMA/LandHydrology.jl/issues/19

File: src/SoilModel/parameters.jl (hunk 3 above)
Comment (kmdeck) [reply src/SoilModel/parameters.jl:5]: https://github.com/CliMA/LandHydrology.jl/issues/20
